src/Course/State.hs

```diff
@@ -191,8 +271,39 @@ distinct =
 --
 -- >>> isHappy 44
 -- True
-isHappy ::
-  Integer
-  -> Bool
-isHappy =
-  error "todo: Course.State#isHappy"
+
+-- attempt 1: NOT using firstRepeat, produce, join, contains
+--
+-- isHappy :: Integer -> Bool
+-- isHappy n = isHappy1 n S.empty
+--   where
+--     isHappy1 :: Integer -> S.Set Integer -> Bool
+--     isHappy1 1 _ = True
+--     isHappy1 m set
+--       | S.member m set = False
+--       | otherwise = isHappy1 (sumOfSquareOfDigits m) (S.insert m set)
+--     sumOfSquareOfDigits :: Integer -> Integer
+--     sumOfSquareOfDigits m = sumIntegers (map square (digits m))
+--     sumIntegers :: List Integer -> Integer
+--     sumIntegers ms = foldRight (+) 0 ms
+--     square :: Integer -> Integer
+--     square m = m * m
+--     digits :: Integer -> List Integer
+--     digits m = map charToInteger (listh (show m))
+--     charToInteger :: Char -> Integer
+--     charToInteger = toInteger . Data.Char.digitToInt
+
+-- attempt 2: using firstRepeat, produce, join, contains
+isHappy :: Integer -> Bool
+isHappy n = contains 1 $ firstRepeat (produce sumOfSquareOfDigits n)
+  where
+    sumOfSquareOfDigits :: Integer -> Integer
+    sumOfSquareOfDigits m = sumIntegers (map square (digits m))
+    sumIntegers :: List Integer -> Integer
+    sumIntegers ms = foldRight (+) 0 ms
+    square :: Integer -> Integer
+    square = join (*)
```

Comment:
wasn't that one a pain? guh... took me *ages* to work out `join (*)` was the `((->) r)` instance applied to itself. lol. What a head-bang.